libdrgn/dwarf_info.c

```diff
@@ -5336,6 +5336,77 @@ drgn_object_from_dwarf(struct drgn_debug_info *dbinfo,
 					       function_die, regs, ret);
 }
 
+DEFINE_VECTOR(drgn_char_p_vector, const char *);
+
+static struct drgn_error *add_dwarf_enumerators(Dwarf_Die *enumeration_type,
+						struct drgn_char_p_vector *vec)
+{
+	Dwarf_Die child;
+	int r = dwarf_child(enumeration_type, &child);
+	while (r == 0) {
+		if (dwarf_tag(&child) == DW_TAG_enumerator) {
+			const char *name = dwarf_diename(&child);
+			if (!drgn_char_p_vector_append(vec, &name))
+				return &drgn_enomem;
+		}
+		r = dwarf_siblingof(&child, &child);
+	}
+	if (r < 0)
+		return drgn_error_libdw();
+	return NULL;
+}
+
+struct drgn_error *drgn_dwarf_scopes_names(Dwarf_Die *scopes,
+					   size_t num_scopes,
+					   const char ***names_ret,
+					   size_t *count_ret)
+{
+	Dwarf_Die die;
+	struct drgn_char_p_vector vec = VECTOR_INIT;
+	struct drgn_error *err;
+
+	for (size_t scope = 0; scope < num_scopes; scope++) {
+		const char *name;
+		if (dwarf_child(&scopes[scope], &die) != 0)
+			continue;
+		do {
+			switch (dwarf_tag(&die)) {
+			case DW_TAG_variable:
+			case DW_TAG_formal_parameter:
+			case DW_TAG_subprogram:
+				if (!(name = dwarf_diename(&die)))
+					continue;
+				if (!drgn_char_p_vector_append(&vec, &name)) {
+					err = &drgn_enomem;
+					goto out_err;
+				}
+				break;
+			case DW_TAG_enumeration_type: {
+				bool enum_class;
+				if (dwarf_flag_integrate(&die, DW_AT_enum_class,
+							 &enum_class))
+					return drgn_error_libdw();
```

Comment:
Needs to `goto out_err` instead of `return`.

Reply:
oof 😅